theano/gof/wrapper.py

```diff
@@ -0,0 +1,446 @@
+    (eg. TensorType('int64', (False,))).
+
+    In a Python code any attribute named `key` will be available via:
+        structObject.key
+
+    In a C code, attributes created to represent an instance of the type associated to `key` will be available via:
+        structObject.key
+        structObject.dtype_key # e.g. from TensorType C code.
+        structObject.other_attribute_named_from_key
+        etc.
+
+    """
+
+    def __init__(self, **kwargs):
+        if len(kwargs) == 0:
+            raise ValueError('Cannot create Wrapper from empty data.')
+
+        for attribute_name in kwargs:
+            if re.match('^[A-Za-z_][A-Za-z0-9_]*$', attribute_name) is None:
+                raise SyntaxError('Wrapper: attribute "%s" should be a valid identifier.' % attribute_name)
+            type_instance = kwargs[attribute_name]
+            type_name = type_instance.__class__.__name__
+            if not isinstance(type_instance, Type):
+                raise TypeError('Wrapper: attribute "%s" should inherit from theano Type, got "%s".'
+                                % (attribute_name, type_name))
+
+        self.length = len(kwargs)
+        self.fields = tuple(sorted(kwargs.keys()))
+        self.types = tuple(kwargs[field] for field in self.fields)
+        self.name = self.generate_struct_name()
+
+    def __repr__(self):
+        return 'Wrapper<%s>' % ', '.join([('%s:%s' % (self.fields[i], self.types[i])) for i in range(self.length)])
+
+    def __eq__(self, other):
+        # To be checked.
+        return (type(self) == type(other) and self.fields == other.fields and self.types == other.types)
+
+    def __hash__(self):
+        return hash((type(self),) + self.fields + self.types)
+
+    def generate_struct_name(self):
+        """"
+        This method tries to generate an unique name for the current instance.
+        This name is intended to be used as struct name in C code and as constant
+        definition to check if a similar Wrapper has already been created
+        (see c_support_code() below).
+        """
+        fields_string = ','.join(self.fields).encode('utf-8')
+        types_string = ','.join(str(t) for t in self.types).encode('utf-8')
+        fields_hex = hashlib.md5(fields_string).hexdigest()
+        types_hex = hashlib.md5(types_string).hexdigest()
+        return '_wrapper_struct_%s_%s' % (fields_hex, types_hex)
+
+    def check_that_values_are_compatible(self, data, strict, allow_downcast):
+        wrap_instance = dict()
+        for i in range(self.length):
+            wrap_instance[self.fields[i]] = self.types[i].filter(getattr(data, self.fields[i]), strict, allow_downcast)
+        return Wrap(**wrap_instance)
+
+    # Returns a wrapped object with expected attributes or (in strict mode) checks that data has expected attributes.
+    def filter(self, data, strict=False, allow_downcast=None):
+        if strict:
```

Comment:
In that case, probably not.  However, please try to avoid three times what is essentially the same code.

You could have a if not strict section first that converts the dict to a Wrap and then check the properties.